code/controllers/subsystem/persistence.dm

```diff
@@ -32,55 +39,156 @@ SUBSYSTEM_DEF(persistence)
-	var/list/saved_messages = list()
-	if(fexists("data/npc_saves/ChiselMessages.sav")) //legacy compatability to convert old format to new
-		var/savefile/chisel_messages_sav = new /savefile("data/npc_saves/ChiselMessages.sav")
-		var/saved_json
-		chisel_messages_sav[SSmapping.config.map_name] >> saved_json
-		if(!saved_json)
-			return
-		saved_messages = json_decode(saved_json)
-		fdel("data/npc_saves/ChiselMessages.sav")
-	else
-		var/json_file = file("data/npc_saves/ChiselMessages[SSmapping.config.map_name].json")
-		if(!fexists(json_file))
-			return
-		var/list/json = json_decode(file2text(json_file))
+/datum/controller/subsystem/persistence/proc/load_wall_engravings()
+	var/json_file = file(ENGRAVING_SAVE_FILE)
+	if(!fexists(json_file))
+		return
+	var/list/json = json_decode(file2text(json_file))
+	if(!json)
+		return
 
-		if(!json)
-			return
-		saved_messages = json["data"]
+	if(json["version"] < ENGRAVING_PERSISTENCE_VERSION)
+		update_wall_engravings(json)
 
-	for(var/item in saved_messages)
-		if(!islist(item))
-			continue
+	var/successfully_loaded_engravings = 0
 
-		var/xvar = item["x"]
-		var/yvar = item["y"]
-		var/zvar = item["z"]
+	var/list/viable_turfs = get_area_turfs(/area/maintenance) + get_area_turfs(/area/security/prison)
+	var/list/turfs_to_pick_from = list()
 
-		if(!xvar || !yvar || !zvar)
+	for(var/turf/T as anything in viable_turfs)
+		if(!isclosedturf(T))
 			continue
+		turfs_to_pick_from += T
 
-		var/turf/T = locate(xvar, yvar, zvar)
-		if(!isturf(T))
-			continue
+	var/list/engraving_entries = json["entries"]
+
+	if(engraving_entries.len)
+		for(var/engraving_index in MIN_PERSISTENT_ENGRAVINGS to MAX_PERSISTENT_ENGRAVINGS)
+			var/engraving = engraving_entries[rand(1, engraving_entries.len)] //This means repeats will happen for now, but its something I can live with. Just make more engravings!
+			if(!islist(engraving))
+				stack_trace("something's wrong with the engraving data! one of the saved engravings wasn't a list!")
+				continue
+
+			var/turf/closed/engraved_wall = pick(turfs_to_pick_from)
+
+			if(HAS_TRAIT(engraved_wall, NOT_ENGRAVABLE))
+				continue
+
+			engraved_wall.AddComponent(/datum/component/engraved, engraving["story"], FALSE, engraving["story_value"])
+			successfully_loaded_engravings++
+			turfs_to_pick_from -= engraved_wall
+
+	log_world("Loaded [successfully_loaded_engravings] engraved messages on map [SSmapping.config.map_name]")
+
+/datum/controller/subsystem/persistence/proc/save_wall_engravings()
+	var/list/saved_data = list()
 
-		if(locate(/obj/structure/chisel_message) in T)
+	saved_data["version"] = ENGRAVING_PERSISTENCE_VERSION
+	saved_data["entries"] = list()
+
+
+	var/json_file = file(ENGRAVING_SAVE_FILE)
+	if(fexists(json_file))
+		var/list/old_json = json_decode(file2text(json_file))
+		if(old_json)
+			saved_data["entries"] = old_json["entries"]
+
+	for(var/datum/component/engraved/engraving in wall_engravings)
```

Comment:
Shouldn't this be running some dedupe function on the resultant json/list ?

Reply:
save paintings doesn't:
```dm
	for(var/obj/structure/sign/painting/P in painting_frames)
		P.save_persistent()

	var/json_file = file("data/paintings.json")
	fdel(json_file)
	WRITE_FILE(json_file, json_encode(paintings))
```

Reply:
@AnturK When would a dupe occur?